### PR TITLE
Added autodetect_encoding option to csvLoader

### PR DESCRIPTION
**Pull request by Taikono-Himazin**
Added autodetect_encoding option to csvLoader like TextLoader.

### CODE REVIEW

File: libs/langchain/langchain/document_loaders/csv_loader.py [hunk 1]
```diff
@@ -46,17 +48,21 @@ def __init__(
...
+        self.autodetect_encoding = autodetect_encoding
 
     def load(self) -> List[Document]:
         """Load data into document objects."""
 
         docs = []
-        with open(self.file_path, newline="", encoding=self.encoding) as csvfile:
+        if self.autodetect_encoding:
+            detected_encodings = detect_file_encodings(self.file_path)
+        with open(self.file_path, newline="", encoding=detected_encodings if self.autodetect_encoding else self.encoding) as csvfile:
```
Comment (efriis) [inline libs/langchain/langchain/document_loaders/csv_loader.py:65]: This has different behavior than `TextLoader` in that even if an encoding is defined, the autodetected one would by mistake take precedence

**Review by efriis — COMMENTED** [review]
Hey @Taikono-Himazin - thanks for contributing! Could you 

- change the logic in `load` to be a bit more like [`TextLoader`](https://github.com/langchain-ai/langchain/blob/6e848b879ad551e930462c71e174be0cc730b4c9/libs/langchain/langchain/document_loaders/text.py#L40-L56)?
- add a unit test [like the TextLoader](https://github.com/langchain-ai/langchain/blob/6e848b879ad551e930462c71e174be0cc730b4c9/libs/langchain/tests/unit_tests/document_loaders/test_detect_encoding.py#L10-L26)

File: libs/langchain/langchain/document_loaders/csv_loader.py (hunk 1 above)
Comment (Taikono-Himazin) [reply libs/langchain/langchain/document_loaders/csv_loader.py]: I changed the code.
Could you please review it?

**Comment by efriis** [comment]
Great changes! Last but not least, could you `make format` to get lint passing and get the new unit test passing?

**Comment by Taikono-Himazin** [comment]
I did make format.
Is there anything else to do?

**Comment by efriis** [comment]
@Taikono-Himazin thanks! I would guess the unit test you added will still fail give it was just a `make format` change. Here's a link to the one that failed yesterday: https://github.com/langchain-ai/langchain/actions/runs/6406293897/job/17406969053

You can also iterate on them locally with
```
cd libs/langchain

# install required deps (only needed the first time)
poetry install -E extended_testing

make extended_tests
```

**Comment by Taikono-Himazin** [comment]
Oh, I'm sorry.
I didn't really know how to run tests.
I modified the code and made it pass the test.
Sorry to bother you, but could you please check again?

**Comment by Taikono-Himazin** [comment]
タイプチェックで失敗していますね。確認して修正します。

**Comment by Taikono-Himazin** [comment]
Oh, it's failing the type check. I'll check and fix it.

**Comment by Taikono-Himazin** [comment]
```
Argument "loader_cls" to "DirectoryLoader" has incompatible type "Type[CSVLoader]"; expected "Union[Type[UnstructuredFileLoader], Type[TextLoader], Type[BSHTMLLoader]]
```
I got an error like the one above.
But I don't see anything wrong with this.
Isn't DirectoryLoader's incompatibleType wrong?
Shouldn't it be BaseLoader instead of UnstructuredFileLoader?

**Comment by Taikono-Himazin** [comment]
I fixed it. Could you please confirm?

**Review by efriis — APPROVED** [review]
Looks great thanks!

**Pull request merged**